cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/functional/BasicLattices.kt

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2024, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.helpers.functional
+
+import de.fraunhofer.aisec.cpg.helpers.IdentitySet
+import de.fraunhofer.aisec.cpg.helpers.identitySetOf
+import de.fraunhofer.aisec.cpg.helpers.toIdentitySet
+import java.util.IdentityHashMap
+import kotlin.Pair
+import kotlin.collections.component1
+import kotlin.collections.component2
+import kotlin.collections.toMap
+
+/**
+ * A complete lattice is an ordered structure of values of type [T]. [T] could be anything, e.g., a
+ * set, a new data structure (like a range), or anything else. [T] depends on the analysis and
+ * typically has to abstract the value for the specific purpose.
+ *
+ * This class is actually used to hold individual instances of the lattice's elements and to compute
+ * bigger elements depending on these two elements.
+ *
+ * Implementations of this class have to implement the comparator, the least upper bound of two
+ * lattices.
+ */
+interface LatticeElement<T> : Comparable<LatticeElement<T>> {
+    val elements: T
+
+    /**
+     * Computes the least upper bound of this lattice and [other]. It returns a new object and does
+     * not modify either of the objects.
+     */
+    fun lub(other: LatticeElement<out T>): LatticeElement<in T>
+
+    /** Duplicates the object, i.e., makes a deep copy. */
+    fun duplicate(): LatticeElement<in T>
+}
+
+typealias PowersetLatticeT<V> = PowersetLattice<IdentitySet<V>, V>
+
+inline fun <reified V> emptyPowersetLattice() = PowersetLattice<IdentitySet<V>, V>(identitySetOf())
+
+/**
+ * Implements the [LatticeElement] for a lattice over a set of nodes. The lattice itself is
+ * constructed by the powerset.
+ */
+open class PowersetLattice<V : IdentitySet<T>, T>(override val elements: V) : LatticeElement<V> {
```

Comment:
I fail to understand what this class actually makes a powerset lattice.

Reply:
If I understand it correctly, then *if* the elements would contain a powerset, then this would be considered a lattice because then the lab is provided by set union.

BTW: why don't we need the greatest lower bound?

Reply:
I renamed it to `PowersetLatticeElement` because it implements only a single element (i.e., one set and the operations).

A greatest lower bound is required for a complete lattice which we do not necessarily need here